File: mandelbrot.R
# Build instructions for shared library
#
# Linux and Windows:
# gfortran -c -fPIC mandelbrot.f90
# gfortran -shared -o mandelbrot.so mandelbrot.o
#
# AIX:
# gfortran -maix64 -c -fPIC mandelbrot.f90
# gfortran -maix64 -static-libgcc -shared -o mandelbrot.so mandelbrot.o
# Add underscore to symbol name

# Blue colour palette
blue <- colorRampPalette(c("#000000", "#0000FF"))

# Implementation of Mandelbrot using vectorised R
mandel_R <- function(nx, ny, maxiter, window, cut, magnitude) {

  # Set axes
  xaxis <- seq(window[1], window[3], length.out=nx)
  yaxis <- seq(window[2], window[4], length.out=ny)

  # Initial values for mandelbrot set
  C <- matrix(rep(xaxis, ny), nx, ny, byrow=FALSE)
  C <- C + 1i*matrix(rep(yaxis, ny), nx, ny, byrow=TRUE)

  # Iteration, only compute elements below cutoff
  Z <- matrix(as.complex(0), nx, ny)
  for (k in 1:maxiter) {
    idx <- (abs(Z) <= cut)
    if (sum(idx) == 0) {break}
    Z[idx] <- Z[idx]*Z[idx]+C[idx]
  }
  magnitude <- abs(Z)
  return(magnitude)
}

# Implementation of Mandelbrot using Fortran
mandel_fortran <- function(nx, ny, maxiter, window, cut, magnitude) {
  result <- .Fortran("mandelbrot", nx = as.integer(nx), ny = as.integer(ny),
                    maxiter = as.integer(maxiter), window = as.double(window),
                    cut = as.double(cut), magnitude = as.double(magnitude))
  return(result$magnitude)
}

# Set parameters
nx <- 800
ny <- 800
maxiter <- 100
cut <- 4.0
window <- c(-2.2, -1.2, 1.0, 1.2)
magnitude <- matrix(0.0, nx, ny)

# Run R version and plot result
starttime <- as.numeric(Sys.time())
magnitude_R <- mandel_R(nx, ny, maxiter, window, cut, magnitude)
stoptime <- as.numeric(Sys.time())
print(paste("R version:", stoptime-starttime))
image(matrix(magnitude_R, nx, ny), col = blue(256))
title("Mandelbrot - R version")

# Run Fortran library version and plot result
dyn.load("mandelbrot.so")
starttime <- as.numeric(Sys.time())
magnitude_fortran <- mandel_fortran(nx, ny, maxiter, window, cut, magnitude)
stoptime <- as.numeric(Sys.time())
print(paste("Fortran version:", stoptime-starttime))
image(matrix(magnitude_fortran, nx, ny), col = blue(256))
title("Mandelbrot - Fortran library version")
dyn.unload("mandelbrot.so")

File: mandelbrot.f90
SUBROUTINE mandelbrot(nx, ny, maxiter, window, cut, magnitude)

    ! Get kind parameters for long integers and double precision
    USE ISO_FORTRAN_ENV, ONLY: INT32, REAL64

    IMPLICIT NONE

    ! Dummy arguments
    INTEGER(kind=INT32), INTENT(in) :: nx, ny, maxiter
    REAL(kind=REAL64), INTENT(in) :: window(4), cut
    REAL(kind=REAL64), INTENT(out) :: magnitude(nx, ny)

    ! Work variables
    INTEGER(kind=INT32) :: i, j, k
    REAL(kind=REAL64) :: start, span, xaxis(nx), yaxis(ny), abs_z
    COMPLEX(kind=REAL64) :: c, z

    ! Set axes, requires type conversion integer to real
    start = window(1)
    span = window(3) - window(1)
    xaxis = start + span*REAL((/(i, i = 1, nx)/)-1, kind=REAL64)/REAL(nx-1, kind=REAL64)

    start = window(2)
    span = window(4) - window(2)
    yaxis = start + span*REAL((/(i, i = 1, ny)/)-1, kind=REAL64)/REAL(ny-1, kind=REAL64)

    ! Compute Mandelbrot function for each element of the set and store its magnitude
    DO j = 1, ny
        DO i = 1, nx
           c = CMPLX(xaxis(i), yaxis(j))
           z = CMPLX(0.0_REAL64, 0.0_REAL64)
           DO k = 1, maxiter
              abs_z = ABS(z)
              IF (abs_z > cut) EXIT
              z = z*z+c
           END DO
           magnitude(i,j) = abs_z
        END DO
    END DO

END SUBROUTINE mandelbrot